File: llang/L0/lib/ast.ml
(* This module contains the abstract syntax tree for the language *)

open Symbols

type l_type = 
  | TInt
  | TBool

type bin_op =   
  | Add
  | Sub
  | Mul
  | Div
  | And
  | Or
  | Eq
  | Ne
  | Lt
  | Le
  | Gt
  | Ge

type un_op = 
  | Neg
  | Not

type ast = 
  | Num of int
  | Bool of bool
  | BinOp of bin_op * ast * ast
  | UnOp of un_op * ast

File: llang/L0/rust/src/ast.rs
use std::rc::Rc;

pub type Type = Rc<TypeData>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TypeData {
    Int,
    Bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    And,
    Or,
    Eq,
    Lt,
    Gt,
    Ne,
    Le,
    Ge,
}

pub type Symbol = i64;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum UnOp {
    Neg,
    Not
}

pub type Ast = Rc<AstData>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum AstData {
    Num(i64),
    Bool(bool),
    BinOp(BinOp, Ast, Ast),
    UnOp(UnOp, Ast),
}

